File: blue-core/src/main/java/blue/noteProcessor/PchInversionProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;

/**
 * Title: blue Description: an object composition environment for csound
 * Copyright: Copyright (c) 2001 Company: steven yi music
 * 
 * @author steven yi
 * @version 1.0
 */

@NoteProcessorPlugin(displayName="PchInversionProcessor", position = 100)
public class PchInversionProcessor implements NoteProcessor {

    double value = 8.00f;
    int pfield = 4;

    public PchInversionProcessor() {
    }
    public PchInversionProcessor(PchInversionProcessor pip) {
        value = pip.value;
        pfield = pip.pfield;
    }

    @Override
    public String toString() {
        // return "[add] pfield: " + pfield + " value: " + value;
        return "[pch inversion]";
    }

    public String getPfield() {
        return Integer.toString(pfield);
    }

    public void setPfield(String pfield) {
        this.pfield = Integer.parseInt(pfield);
    }

    public String getVal() {
        return Double.toString(value);
    }

    public void setVal(String value) {
        this.value = Double.parseDouble(value);
    }

    @Override
    public final NoteList processNotes(NoteList in) throws NoteProcessorException {
        Note temp;
        String val;
        for (int i = 0; i < in.size(); i++) {
            temp = in.get(i);

            try {
                val = temp.getPField(pfield).trim();
                Double.parseDouble(val);
            } catch (NumberFormatException ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.pfieldNotDouble"),
                        pfield);
            } catch (Exception ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield);
            }

            double baseTen = blue.utility.ScoreUtilities.getBaseTen(val);
            double baseTenAxis = blue.utility.ScoreUtilities.getBaseTen(this
                    .getVal());

            double addVal = -1 * (baseTen - baseTenAxis);

            baseTen = baseTenAxis + addVal;

            int octave = (int) (baseTen / 12);
            double strPch = (baseTen % 12) / 100;

            temp.setPField(Double.toString(octave + strPch), pfield);
        }
        return in;
    }

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + i + " 2 6.0" + i + " 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        PchInversionProcessor ap = new PchInversionProcessor();
        ap.setPfield("4");
        ap.setVal("5.00");
        try {
            ap.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");
    }

    public static NoteProcessor loadFromXML(Element data) {
        PchInversionProcessor pip = new PchInversionProcessor();

        pip.setPfield(data.getElement("pfield").getTextString());
        pip.setVal(data.getElement("value").getTextString());

        return pip;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    @Override
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("pfield").setText(this.getPfield());
        retVal.addElement("value").setText(this.getVal());

        return retVal;
    }

    @Override
    public PchInversionProcessor deepCopy() {
        return new PchInversionProcessor(this);
    }
}

File: packages/blue-data/src/note-processors/pch-inversion-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';
import { getBaseTen } from '../utilities/score';

const JAVA_TYPE = 'blue.noteProcessor.PchInversionProcessor';

export class PchInversionProcessor extends NoteProcessor {
  private _value = 8.0;
  private _pfield = 4;

  constructor();
  constructor(src: PchInversionProcessor);
  constructor(src?: PchInversionProcessor) {
    super();
    if (src) {
      this._value = src._value;
      this._pfield = src._pfield;
    }
  }

  getPfield(): string {
    return this._pfield.toString();
  }
  setPfield(pfield: string): void {
    this._pfield = parseInt(pfield, 10);
  }

  getVal(): string {
    return this._value.toString();
  }
  setVal(value: string): void {
    this._value = parseFloat(value);
  }

  override process(notes: NoteList): NoteList {
    for (let i = 0; i < notes.length; i++) {
      const temp = notes.getNote(i);
      let val: string;
      try {
        val = temp.getPField(this._pfield)!.trim();
        parseFloat(val);
      } catch {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      if (isNaN(parseFloat(val))) {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }

      const baseTen = getBaseTen(val);
      const baseTenAxis = getBaseTen(this.getVal());

      const addVal = -1 * (baseTen - baseTenAxis);
      const result = baseTenAxis + addVal;

      const octave = Math.trunc(result / 12);
      const strPch = (result % 12) / 100;

      temp.setPField((octave + strPch).toString(), this._pfield);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'PchInversionProcessor';
  }

  override deepCopy(): PchInversionProcessor {
    return new PchInversionProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('pfield').setText(this.getPfield());
    elem.addElement('value').setText(this.getVal());
    return elem;
  }

  static loadFromXML(data: Element): PchInversionProcessor {
    const proc = new PchInversionProcessor();
    const pf = data.getTextString('pfield');
    if (pf !== null) proc._pfield = parseInt(pf, 10);
    const v = data.getTextString('value');
    if (v !== null) proc._value = parseFloat(v);
    return proc;
  }
}
